### PR TITLE
perfschema: fixed memory leak

### PR DESCRIPTION
**Pull request by nieyy**
For issue https://github.com/pingcap/tidb/issues/1037

### CODE REVIEW

**Comment by nieyy** [comment]
PTAL, @shenli @coocood

File: table/tables/bounded_tables.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+// Seek seeks the handle
+func (t *BoundedTable) Seek(ctx context.Context, handle int64) (int64, bool, error) {
+	// no lock needed during seek since we don't care occasionally data
+	// errors.
+	result := int32(-1)
+	if handle < invalidRecordID {
+		// this is the first seek call
+		result = 0
+	} else {
+		for i, record := range t.records {
```
Comment (shenli) [inline table/tables/bounded_tables.go]: Can we use a map to improve performance?

Comment (nieyy) [inline table/tables/bounded_tables.go]: `Seek` used for `SELECT` operation only, it's not related to running performance, also, the array is not too big, so I don't think we need to use a map.

File: perfschema/init.go [hunk 2]
```diff
@@ -193,15 +202,14 @@ func createMemoryTable(meta *model.TableInfo, alloc autoid.Allocator) (table.Tab
 	return tbl, nil
 }
 
+func createBoundedTable(meta *model.TableInfo, alloc autoid.Allocator, capacity int32) (table.Table, error) {
+	tbl, _ := tables.BoundedTableFromMeta(alloc, meta, capacity)
```
Comment (c4pt0r) [inline perfschema/init.go]: why not checking error?

File: table/tables/bounded_tables.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+		col := &column.Col{ColumnInfo: *colInfo}
+		columns = append(columns, col)
+		if col.IsPKHandleColumn(tblInfo) {
+			pkHandleColumn = col
+		}
+	}
+	t := newBoundedTable(tblInfo.ID, tblInfo.Name.O, columns, alloc, capacity)
+	t.pkHandleCol = pkHandleColumn
+	t.meta = tblInfo
+	return t, nil
```
Comment (c4pt0r) [inline table/tables/bounded_tables.go]: the function returns an error, but the implementation here always return a nil, so is it necessary?

File: perfschema/init.go (hunk 2 above)
Comment (nieyy) [inline perfschema/init.go]: createMemoryTable did not check error either, but anyway, I can revise

File: table/tables/bounded_tables.go (hunk 3 above)
Comment (nieyy) [inline table/tables/bounded_tables.go]: same as `MemoryTableFromMeta`, I can revise if you insist

Comment (coocood) [inline table/tables/bounded_tables.go]: @nieyy we can remove the error in returned value list.

Comment (nieyy) [inline table/tables/bounded_tables.go]: ok,

File: table/tables/bounded_tables.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		Columns:      cols,
+		recordPrefix: genTableRecordPrefix(tableID),
+		records:      make([]boundedItem, capacity),
+		capacity:     capacity,
+		cursor:       0,
+	}
+	return t
+}
+
+// Seek seeks the handle
```
Comment (zimulala) [inline table/tables/bounded_tables.go]: add . at the end of line

File: table/tables/bounded_tables.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+func (t *BoundedTable) Seek(ctx context.Context, handle int64) (int64, bool, error) {
+	// no lock needed during seek since we don't care occasionally data
+	// errors.
+	result := int32(-1)
+	if handle < invalidRecordID {
```
Comment (zimulala) [inline table/tables/bounded_tables.go]: invalidRecordID -> minRecordID ? 
Less than the invalidRecordID isn't necessarily invalid.

File: table/tables/bounded_tables.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta/autoid"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/types"
+)
+
+const (
+	initialRecordID int64 = 123
```
Comment (zimulala) [inline table/tables/bounded_tables.go]: Add some comments to explain why the initialRecordID  is 123 ?

File: table/tables/bounded_tables.go (hunk 4 above)
Comment (nieyy) [inline table/tables/bounded_tables.go]: ok

File: table/tables/bounded_tables.go (hunk 5 above)
Comment (nieyy) [inline table/tables/bounded_tables.go]: Actually the first `Seek` call will use a negative `handle` value, I will change to 0.

File: table/tables/bounded_tables.go (hunk 6 above)
Comment (nieyy) [inline table/tables/bounded_tables.go]: ok

**Comment by nieyy** [comment]
All comments have been revised, PTAL ag, @coocood @shenli @zimulala @c4pt0r

File: perfschema/init.go [hunk 7]
```diff
@@ -32,6 +32,15 @@ type columnInfo struct {
 	elems []string
 }
 
+const (
+	// Maximu concurrent number of elements in table events_xxx_current.
```
Comment (shenli) [inline perfschema/init.go]: Maximu -> Maximum

Comment (nieyy) [inline perfschema/init.go]: ok

File: perfschema/init.go [hunk 8]
```diff
@@ -193,15 +202,14 @@ func createMemoryTable(meta *model.TableInfo, alloc autoid.Allocator) (table.Tab
 	return tbl, nil
 }
 
+func createBoundedTable(meta *model.TableInfo, alloc autoid.Allocator, capacity int32) table.Table {
+	tbl := tables.BoundedTableFromMeta(alloc, meta, capacity)
```
Comment (zimulala) [inline perfschema/init.go]: Return it directly ?

Comment (zimulala) [inline perfschema/init.go]: Return it directly ?

Comment (nieyy) [inline perfschema/init.go]: ok

File: perfschema/init.go [hunk 9]
```diff
@@ -210,9 +218,19 @@ func (ps *perfSchema) buildTables() error {
...
-			return errors.Trace(err)
+
+		var tbl table.Table
+		var err error
+		switch name {
+		case TableStmtsCurrent, TablePreparedStmtsInstances, TableTransCurrent, TableStagesCurrent:
+			tbl = createBoundedTable(meta, alloc, currentElemMax)
+		case TableStmtsHistory, TableStmtsHistoryLong, TableTransHistory, TableTransHistoryLong, TableStagesHistory, TableStagesHistoryLong:
+			tbl = createBoundedTable(meta, alloc, historyElemMax)
+		default:
```
Comment (shenli) [inline perfschema/init.go]: Why not all tables in ps use BoundedTable?

File: perfschema/init.go [hunk 10]
```diff
@@ -210,9 +218,19 @@ func (ps *perfSchema) buildTables() error {
 			c.ID = autoid.GenLocalSchemaID()
 		}
 		alloc := autoid.NewMemoryAllocator(dbID)
-		tbl, err := createMemoryTable(meta, alloc)
-		if err != nil {
-			return errors.Trace(err)
+
+		var tbl table.Table
+		var err error
```
Comment (zimulala) [inline perfschema/init.go]: Remove this line, it isn't necessary.

File: perfschema/init.go (hunk 9 above)
Comment (nieyy) [inline perfschema/init.go]: `setup_xxx` tables are used to store configuration information, it's similar to other schema tables, so I would like these tables to use MemoryTable.

File: perfschema/init.go (hunk 10 above)
Comment (nieyy) [inline perfschema/init.go]: ok

File: table/tables/bounded_tables_test.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	tblInfo := &model.TableInfo{
+		ID:      100,
+		Name:    model.NewCIStr("t"),
+		Columns: []*model.ColumnInfo{col1, col2},
+	}
+	alloc := autoid.NewMemoryAllocator(int64(10))
+	ts.tbl = tables.BoundedTableFromMeta(alloc, tblInfo, 1024)
+}
+
+func (ts *testBoundedTableSuite) TestBoundedBasic(c *C) {
```
Comment (zimulala) [inline table/tables/bounded_tables_test.go]: Add `defer testleak.AfterTest(c)()`  to each test.

**Comment by shenli** [comment]
Rest LGTM

File: table/tables/bounded_tables_test.go (hunk 11 above)
Comment (nieyy) [inline table/tables/bounded_tables_test.go]: ok

**Comment by nieyy** [comment]
PTAL ag, @zimulala @coocood

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
Unit tests without Seek and UpdateRecord ?

**Comment by nieyy** [comment]
@zimulala , SELECT operations will call `Seek`, also, user UPDATE operations are not supported for `events_xxx` tables.

**Comment by zimulala** [comment]
LGTM

**Comment by ngaut** [comment]
This branch is out-of-date, Could you rebase master ? @nieyy

**Comment by coocood** [comment]
LGTM

**Comment by nieyy** [comment]
sure, @ngaut

**Comment by nieyy** [comment]
@ngaut , squashed and rebased

**Pull request merged**